File: python/matp/groups/fractal_ratchet.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Fractal Group Ratchet - Matryoshka's Group Encryption Algorithm

Original design for multi-party encrypted communication.
NOT based on Megolm or any existing protocol - clean room implementation.

License: Apache 2.0
Author: Sangeet Sharma
"""

import secrets
import hashlib
import base64
import json
import time
from cryptography.hazmat.primitives.ciphers.aead import AESGCM
from cryptography.hazmat.primitives.kdf.hkdf import HKDF
from cryptography.hazmat.primitives import hashes


class FractalGroupRatchet:
    """
    Fractal Group Ratchet - Efficient group encryption with forward secrecy.
    
    Design Philosophy:
    - Like Russian dolls: Each message encrypted in layers
    - Fractal tree structure: Keys derived from single seed
    - Forward secrecy: Each message uses unique key
    - Efficient: One encryption, many recipients
    
    Security Properties:
    - Confidentiality: AES-256-GCM
    - Authentication: GCM tag
    - Forward secrecy: Per-message keys
    - Group membership: Seed-based access control
    """
    
    VERSION = "1.0.0"
    ALGORITHM = "fractal-group-ratchet-v1"
    # Use fixed, domain-separating salts for HKDF for better cryptographic hygiene
    LAYER_KEY_SALT = b"matp-fractal-layer-key-salt"
    SEED_ROTATION_SALT = b"matp-fractal-seed-rotation-salt"
    
    def __init__(self, group_seed: bytes = None):
        """
        Initialize Fractal Group Ratchet.
        
        Args:
            group_seed: 32-byte seed for group (shared by all members)
                       If None, generates random seed
        """
        if group_seed is None:
            self.group_seed = secrets.token_bytes(32)
        elif isinstance(group_seed, bytes) and len(group_seed) == 32:
            self.group_seed = group_seed
        else:
            raise ValueError("Group seed must be 32 bytes")
        
        self.message_counter = 0
        self.seed_fingerprint = self._compute_fingerprint()
    
    def _compute_fingerprint(self) -> str:
        """Compute fingerprint of group seed for verification."""
        return hashlib.sha256(self.group_seed).hexdigest()[:16]
    
    def _derive_layer_key(self, layer_index: int) -> bytes:
        """
        Derive encryption key for specific message layer.
        
        Uses HKDF with layer index as info parameter.
        Each layer gets unique key derived from group seed.
        
        Args:
            layer_index: Message index (0, 1, 2, ...)
        
        Returns:
            32-byte encryption key for this layer
        """
        return HKDF(
            algorithm=hashes.SHA256(),
            length=32,
            salt=self.LAYER_KEY_SALT,
            info=f"fractal-layer-{layer_index}".encode('utf-8')
        ).derive(self.group_seed)
    
    def encrypt_for_group(self, plaintext: str) -> dict:
        """
        Encrypt message for entire group.
        
        All group members can decrypt using the same group seed.
        Each message uses a unique key (forward secrecy).
        
        Args:
            plaintext: Message to encrypt
        
        Returns:
            dict: Encrypted envelope with:
                - layer: Message index
                - nonce: Random nonce for AES-GCM
                - ciphertext: Encrypted message
                - seed_fingerprint: Verify correct group
                - timestamp: When encrypted
        """
        # Derive unique key for this message
        layer_key = self._derive_layer_key(self.message_counter)
        
        # Encrypt with AES-256-GCM
        cipher = AESGCM(layer_key)
        nonce = secrets.token_bytes(12)
        
        if isinstance(plaintext, str):
            plaintext = plaintext.encode('utf-8')
        
        ciphertext = cipher.encrypt(nonce, plaintext, None)
        
        # Create message envelope
        envelope = {
            "version": self.VERSION,
            "algorithm": self.ALGORITHM,
            "layer": self.message_counter,
            "nonce": base64.b64encode(nonce).decode('utf-8'),
            "ciphertext": base64.b64encode(ciphertext).decode('utf-8'),
            "seed_fingerprint": self.seed_fingerprint,
            "timestamp": time.time()
        }
        
        # Increment counter for next message
        self.message_counter += 1
        
        return envelope
    
    def decrypt_from_group(self, envelope: dict) -> str:
        """
        Decrypt message from group.
        
        Args:
            envelope: Encrypted message envelope
        
        Returns:
            str: Decrypted plaintext
        
        Raises:
            ValueError: If wrong group seed or corrupted message
        """
        # Verify version
        if envelope.get("version") != self.VERSION:
            raise ValueError(f"Unsupported version: {envelope.get('version')}")
        
        # Verify algorithm
        if envelope.get("algorithm") != self.ALGORITHM:
            raise ValueError(f"Unsupported algorithm: {envelope.get('algorithm')}")
        
        # Verify seed fingerprint (are we in the right group?)
        if envelope["seed_fingerprint"] != self.seed_fingerprint:
            raise ValueError("Wrong group seed - cannot decrypt")
        
        # Derive key for this message layer
        layer_key = self._derive_layer_key(envelope["layer"])
        
        # Decrypt with AES-256-GCM
        cipher = AESGCM(layer_key)
        nonce = base64.b64decode(envelope["nonce"])
        ciphertext = base64.b64decode(envelope["ciphertext"])
        
        try:
            plaintext = cipher.decrypt(nonce, ciphertext, None)
            return plaintext.decode('utf-8')
        except Exception as e:
            raise ValueError(f"Decryption failed: {e}")
    
    def export_session(self, from_layer: int = 0) -> dict:
        """
        Export session for new member.
        
        Allows new members to join mid-conversation.
        They can decrypt messages from 'from_layer' onwards.
        
        Args:
            from_layer: Starting message index for new member
        
        Returns:
            dict: Session data to share with new member
        """
        return {
            "version": self.VERSION,
            "algorithm": self.ALGORITHM,
            "group_seed": base64.b64encode(self.group_seed).decode('utf-8'),
            "start_layer": from_layer,
            "seed_fingerprint": self.seed_fingerprint,
            "exported_at": time.time()
        }
    
    def import_session(self, session_data: dict):
        """
        Import session from group admin.
        
        Args:
            session_data: Session exported by admin
        
        Raises:
            ValueError: If incompatible version or algorithm
        """
        # Verify version
        if session_data.get("version") != self.VERSION:
            raise ValueError(f"Incompatible version: {session_data.get('version')}")
        
        # Verify algorithm
        if session_data.get("algorithm") != self.ALGORITHM:
            raise ValueError(f"Incompatible algorithm: {session_data.get('algorithm')}")
        
        # Import seed and state
        self.group_seed = base64.b64decode(session_data["group_seed"])
        self.message_counter = session_data["start_layer"]
        self.seed_fingerprint = self._compute_fingerprint()
        
        # Verify fingerprint matches
        if self.seed_fingerprint != session_data["seed_fingerprint"]:
            raise ValueError("Session fingerprint mismatch")
    
    def rotate_seed(self) -> bytes:
        """
        Rotate group seed for backward secrecy.
        
        After rotation, old messages cannot be decrypted with new seed.
        Use this periodically or when member leaves.
        
        Returns:
            bytes: New group seed (share with remaining members)
        """
        # Derive new seed from old seed
        new_seed = HKDF(
            algorithm=hashes.SHA256(),
            length=32,
            salt=self.SEED_ROTATION_SALT,
            info=b"seed-rotation"
        ).derive(self.group_seed)
        
        # Update state
        old_seed = self.group_seed
        self.group_seed = new_seed
        self.message_counter = 0  # Reset counter
        self.seed_fingerprint = self._compute_fingerprint()
        
        return new_seed
    
    def get_fingerprint(self) -> str:
        """
        Get group seed fingerprint for verification.
        
        Returns:
            str: 16-character hex fingerprint
        """
        return self.seed_fingerprint


# Example usage and tests
if __name__ == "__main__":
    print("=== Fractal Group Ratchet Demo ===\n")
    
    # Create group
    print("1. Alice creates group")
    alice_ratchet = FractalGroupRatchet()
    group_seed = alice_ratchet.group_seed
    print(f"   Group fingerprint: {alice_ratchet.get_fingerprint()}\n")
    
    # Bob and Charlie join with same seed
    print("2. Bob and Charlie join group")
    bob_ratchet = FractalGroupRatchet(group_seed=group_seed)
    charlie_ratchet = FractalGroupRatchet(group_seed=group_seed)
    print(f"   Bob fingerprint: {bob_ratchet.get_fingerprint()}")
    print(f"   Charlie fingerprint: {charlie_ratchet.get_fingerprint()}\n")
    
    # Alice sends message
    print("3. Alice sends message")
    msg1 = alice_ratchet.encrypt_for_group("Hello everyone!")
    print(f"   Encrypted layer: {msg1['layer']}")
    print(f"   Ciphertext: {msg1['ciphertext'][:40]}...\n")
    
    # Bob and Charlie decrypt
    print("4. Bob and Charlie decrypt")
    bob_sees = bob_ratchet.decrypt_from_group(msg1)
    charlie_sees = charlie_ratchet.decrypt_from_group(msg1)
    print(f"   Bob: {bob_sees}")
    print(f"   Charlie: {charlie_sees}\n")
    
    # Multiple messages (forward secrecy)
    print("5. Multiple messages (each uses unique key)")
    for i in range(3):
        msg = alice_ratchet.encrypt_for_group(f"Message {i+2}")
        decrypted = bob_ratchet.decrypt_from_group(msg)
        print(f"   Layer {msg['layer']}: {decrypted}")
    print()
    
    # Export session for new member
    print("6. Dave joins mid-conversation")
    session = alice_ratchet.export_session(from_layer=alice_ratchet.message_counter)
    dave_ratchet = FractalGroupRatchet()
    dave_ratchet.import_session(session)
    print(f"   Dave can decrypt from layer {dave_ratchet.message_counter} onwards\n")
    
    # Dave receives new message
    msg_for_dave = alice_ratchet.encrypt_for_group("Welcome Dave!")
    dave_sees = dave_ratchet.decrypt_from_group(msg_for_dave)
    print(f"   Dave: {dave_sees}\n")
    
    # Seed rotation
    print("7. Seed rotation (backward secrecy)")
    old_fingerprint = alice_ratchet.get_fingerprint()
    new_seed = alice_ratchet.rotate_seed()
    new_fingerprint = alice_ratchet.get_fingerprint()
    print(f"   Old fingerprint: {old_fingerprint}")
    print(f"   New fingerprint: {new_fingerprint}")
    print(f"   Old messages cannot be decrypted with new seed\n")
    
    print("=== Demo Complete ===")

File: rust/core/src/groups/fractal_ratchet.rs
//! # Fractal Group Ratchet
//!
//! Original design for multi-party encrypted communication.
//! Like Russian dolls: Each message encrypted in layers.
//! Fractal tree structure: Keys derived from single seed.

use aes_gcm::{
    aead::{Aead, KeyInit},
    Aes256Gcm, Nonce,
};
use hkdf::Hkdf;
use sha2::Sha256;
use serde::{Serialize, Deserialize};
use rand::Rng;

use super::GroupError;

const VERSION: &str = "1.0.0";
const ALGORITHM: &str = "fractal-group-ratchet-v1";

/// Encrypted message envelope for group communication
#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct GroupEnvelope {
    pub version: String,
    pub algorithm: String,
    pub layer: u32,
    pub nonce: Vec<u8>,
    pub ciphertext: Vec<u8>,
    pub seed_fingerprint: String,
    pub timestamp: f64,
}

/// Fractal Group Ratchet for efficient group encryption
pub struct FractalGroupRatchet {
    group_seed: [u8; 32],
    pub message_counter: u32,
    seed_fingerprint: String,
}

impl FractalGroupRatchet {
    /// Create new ratchet with optional seed
    pub fn new(group_seed: Option<[u8; 32]>) -> Self {
        let seed = group_seed.unwrap_or_else(|| {
            let mut s = [0u8; 32];
            rand::thread_rng().fill(&mut s);
            s
        });
        
        let fingerprint = Self::compute_fingerprint(&seed);
        
        Self {
            group_seed: seed,
            message_counter: 0,
            seed_fingerprint: fingerprint,
        }
    }
    
    /// Compute fingerprint of group seed
    fn compute_fingerprint(seed: &[u8; 32]) -> String {
        use sha2::Digest;
        let hash = Sha256::digest(seed);
        hex::encode(&hash[..16])
    }
    
    /// Derive encryption key for specific message layer
    fn derive_layer_key(&self, layer_index: u32) -> Result<[u8; 32], GroupError> {
        let info = format!("fractal-layer-{}", layer_index);
        let mut key = [0u8; 32];
        
        Hkdf::<Sha256>::new(Some(b"matp-fractal-layer-key-salt"), &self.group_seed)
            .expand(info.as_bytes(), &mut key)
            .map_err(|_| GroupError::EncryptionError("Key derivation failed".to_string()))?;
        
        Ok(key)
    }
    
    /// Encrypt message for entire group
    pub fn encrypt_for_group(&mut self, plaintext: &str) -> Result<GroupEnvelope, GroupError> {
        let layer_key = self.derive_layer_key(self.message_counter)?;
        
        // Encrypt with AES-256-GCM
        let cipher = Aes256Gcm::new_from_slice(&layer_key)
            .map_err(|_| GroupError::EncryptionError("Cipher init failed".to_string()))?;
        
        let nonce_bytes: [u8; 12] = rand::thread_rng().gen();
        let nonce = Nonce::from(nonce_bytes);
        
        let ciphertext = cipher.encrypt(&nonce, plaintext.as_bytes())
            .map_err(|_| GroupError::EncryptionError("Encryption failed".to_string()))?;
        
        let envelope = GroupEnvelope {
            version: VERSION.to_string(),
            algorithm: ALGORITHM.to_string(),
            layer: self.message_counter,
            nonce: nonce_bytes.to_vec(),
            ciphertext,
            seed_fingerprint: self.seed_fingerprint.clone(),
            timestamp: std::time::SystemTime::now()
                .duration_since(std::time::UNIX_EPOCH)
                .unwrap()
                .as_secs_f64(),
        };
        
        self.message_counter += 1;
        Ok(envelope)
    }
    
    /// Decrypt message from group
    pub fn decrypt_from_group(&self, envelope: &GroupEnvelope) -> Result<String, GroupError> {
        // Verify version
        if envelope.version != VERSION {
            return Err(GroupError::DecryptionError(format!("Unsupported version: {}", envelope.version)));
        }
        
        // Verify algorithm
        if envelope.algorithm != ALGORITHM {
            return Err(GroupError::DecryptionError(format!("Unsupported algorithm: {}", envelope.algorithm)));
        }
        
        // Verify seed fingerprint
        if envelope.seed_fingerprint != self.seed_fingerprint {
            return Err(GroupError::DecryptionError("Wrong group seed".to_string()));
        }
        
        // Derive key for this layer
        let layer_key = self.derive_layer_key(envelope.layer)?;
        
        // Decrypt
        let cipher = Aes256Gcm::new_from_slice(&layer_key)
            .map_err(|_| GroupError::DecryptionError("Cipher init failed".to_string()))?;
        
        let nonce_bytes: [u8; 12] = envelope.nonce.as_slice().try_into().unwrap();
        let nonce = Nonce::from(nonce_bytes);
        let plaintext = cipher.decrypt(&nonce, envelope.ciphertext.as_slice())
            .map_err(|_| GroupError::DecryptionError("Decryption failed".to_string()))?;
        
        String::from_utf8(plaintext)
            .map_err(|_| GroupError::DecryptionError("UTF-8 decode failed".to_string()))
    }
    
    /// Export session for new member
    pub fn export_session(&self, from_layer: u32) -> SessionExport {
        SessionExport {
            version: VERSION.to_string(),
            algorithm: ALGORITHM.to_string(),
            group_seed: self.group_seed,
            start_layer: from_layer,
            seed_fingerprint: self.seed_fingerprint.clone(),
            exported_at: std::time::SystemTime::now()
                .duration_since(std::time::UNIX_EPOCH)
                .unwrap()
                .as_secs_f64(),
        }
    }
    
    /// Import session from admin
    pub fn import_session(&mut self, session: &SessionExport) -> Result<(), GroupError> {
        if session.version != VERSION {
            return Err(GroupError::InvalidState(format!("Incompatible version: {}", session.version)));
        }
        
        if session.algorithm != ALGORITHM {
            return Err(GroupError::InvalidState(format!("Incompatible algorithm: {}", session.algorithm)));
        }
        
        self.group_seed = session.group_seed;
        self.message_counter = session.start_layer;
        self.seed_fingerprint = Self::compute_fingerprint(&self.group_seed);
        
        if self.seed_fingerprint != session.seed_fingerprint {
            return Err(GroupError::InvalidState("Session fingerprint mismatch".to_string()));
        }
        
        Ok(())
    }
    
    /// Rotate group seed for backward secrecy
    pub fn rotate_seed(&mut self) -> [u8; 32] {
        let mut new_seed = [0u8; 32];
        Hkdf::<Sha256>::new(Some(b"matp-fractal-seed-rotation-salt"), &self.group_seed)
            .expand(b"seed-rotation", &mut new_seed)
            .expect("HKDF expand failed");
        
        self.group_seed = new_seed;
        self.message_counter = 0;
        self.seed_fingerprint = Self::compute_fingerprint(&self.group_seed);
        
        new_seed
    }
    
    /// Get group seed fingerprint
    pub fn get_fingerprint(&self) -> &str {
        &self.seed_fingerprint
    }
    
    /// Get group seed (for sharing with new members)
    pub fn get_group_seed(&self) -> [u8; 32] {
        self.group_seed
    }
}

/// Session export for new members
#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct SessionExport {
    pub version: String,
    pub algorithm: String,
    pub group_seed: [u8; 32],
    pub start_layer: u32,
    pub seed_fingerprint: String,
    pub exported_at: f64,
}

#[cfg(test)]
mod tests {
    use super::*;
    
    #[test]
    fn test_group_ratchet() {
        let mut alice = FractalGroupRatchet::new(None);
        let seed = alice.get_group_seed();
        let bob = FractalGroupRatchet::new(Some(seed));
        
        assert_eq!(alice.get_fingerprint(), bob.get_fingerprint());
        
        let msg = alice.encrypt_for_group("Hello group!").unwrap();
        let decrypted = bob.decrypt_from_group(&msg).unwrap();
        
        assert_eq!(decrypted, "Hello group!");
    }
    
    #[test]
    fn test_session_export() {
        let mut alice = FractalGroupRatchet::new(None);
        
        // Send some messages
        for i in 0..3 {
            alice.encrypt_for_group(&format!("Message {}", i)).unwrap();
        }
        
        // Export session for new member
        let session = alice.export_session(alice.message_counter);
        
        let mut dave = FractalGroupRatchet::new(None);
        dave.import_session(&session).unwrap();
        
        assert_eq!(dave.get_fingerprint(), alice.get_fingerprint());
    }
    
    #[test]
    fn test_seed_rotation() {
        let mut ratchet = FractalGroupRatchet::new(None);
        let old_fingerprint = ratchet.get_fingerprint().to_string();
        
        ratchet.rotate_seed();
        let new_fingerprint = ratchet.get_fingerprint();
        
        assert_ne!(old_fingerprint, new_fingerprint);
    }
}
